File: playwright-ts/src/pages/CustomerNewPage.ts
import { expect, type Locator, type Page } from '@playwright/test';

/**
 * Page Object for the "New Customer" form.
 *
 * The kind/theme fields are Syncfusion dropdowns: clicking the field opens a
 * popup of options. We select the first available option rather than a specific
 * label, so the journey doesn't depend on which reference rows happen to be seeded.
 */
export class CustomerNewPage {
  readonly page: Page;
  readonly root: Locator;
  readonly name: Locator;
  readonly submit: Locator;

  constructor(page: Page) {
    this.page = page;
    this.root = page.getByTestId('customer.new');
    this.name = page.locator('#name');
    this.submit = page.getByRole('button', { name: 'Create Customer' });
  }

  async goto(): Promise<void> {
    await this.page.goto('/customer/new');
    await expect(this.root).toBeVisible();
  }

  private async pickFirstOption(fieldId: string): Promise<void> {
    // Open via the wrapper (the readonly input intercepts clicks), wait for the
    // popup, then select the first real option by keyboard. Clicking the popup
    // <li> doesn't reliably commit the Syncfusion selection — ArrowDown (off the
    // "-- Select --" placeholder) + Enter does.
    await this.page.locator(`span.e-ddl:has(input#${fieldId})`).click();
    await this.page.locator('.e-popup-open .e-list-item').first().waitFor();
    await this.page.keyboard.press('ArrowDown');
    await this.page.keyboard.press('Enter');
  }

  async createOrganization(name: string): Promise<void> {
    await this.name.fill(name);
    await this.pickFirstOption('kindSysId');
    await this.pickFirstOption('themeSysId');
    await this.submit.click();
  }
}

File: selenium-java/src/test/java/com/maelstromops/qa/pages/CustomerNewPage.java
package com.maelstromops.qa.pages;

import com.maelstromops.qa.support.Config;
import org.openqa.selenium.By;
import org.openqa.selenium.Keys;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.interactions.Actions;

public class CustomerNewPage extends BasePage {

  private final By root = By.cssSelector("[data-testid='customer.new']");
  private final By name = By.id("name");
  private final By submit = By.xpath("//button[normalize-space()='Create Customer']");

  public CustomerNewPage(WebDriver driver) {
    super(driver);
  }

  public CustomerNewPage open() {
    driver.get(Config.baseUrl() + "/customer/new");
    visible(root);
    return this;
  }

  public void createOrganization(String customerName) {
    type(name, customerName);
    pickFirstOption("kindSysId");
    pickFirstOption("themeSysId");
    click(submit);
  }

  private void pickFirstOption(String fieldId) {
    // Open the Syncfusion dropdown via its wrapper (the readonly input intercepts
    // clicks), then commit the first real option with REAL keyboard events:
    // ArrowDown moves off the "-- Select --" placeholder, Enter commits. Selenium's
    // sendKeys dispatches real key events that the widget honours.
    clickable(By.cssSelector("span.e-ddl:has(input#" + fieldId + ")")).click();
    visible(By.cssSelector(".e-popup-open .e-list-item"));
    new Actions(driver).sendKeys(Keys.ARROW_DOWN, Keys.ENTER).perform();
  }
}
